### PR TITLE
Proposal: Subnet Scarcity

### PR DESCRIPTION
**Pull request by rbtr**
**Reason for Change**:

This starts the design train for the Subnet Scarcity feature, a solution to the issue raised in https://github.com/Azure/azure-container-networking/issues/1605. 

Namely: in SWIFT, due to the batch-wise scaling of the IP Pool, CNS reserves an overhead of IPs from the Subnet on every Node. This can lead to artificial Subnet IP Exhaustion where there are insufficient unreserved IPs left in the Subnet for Nodes to join the cluster or for Pods to schedule, even when the real total Pod IP usage is less than the total Subnet IP Capacity.

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

**Notes**:

### CODE REVIEW

File: docs/feature/subnet-scarcity/phase-1/1-subnetstate.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        '200':
+          description: The matching SubnetState
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/SubnetState'
+        '400':
+          description: bad input parameter
```
Comment (rbtr) [inline docs/feature/subnet-scarcity/phase-1/1-subnetstate.md:37]: @rsagasthya can you fill in the missing response types and codes of this API as it has been written?

Comment (rbtr) [inline docs/feature/subnet-scarcity/phase-1/1-subnetstate.md:37]: @rsagasthya can you fill in the missing response types and codes of this API as it has been written?

File: docs/feature/subnet-scarcity/proposal.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+When a Pod is created, the CNI will call with a request to assign an IP. If CNS is out of IPs and cannot honor that request, the CNI will return an error to the CRI, which will follow up by tearing down that Pod sandbox and starting over. Because of this stateless retrying, CNS can only reliable understand that it needs _at least one more_ IP, because it is impossible to tell if subsequent requests are retries for the same Pod, or many different Pods. If _many_ Pods have been scheduled, CNS will still only request a single additional batch of IPs, and assign those IPs one at a time until it runs out, then request a single additional batch of IPs...
+
+A more predictive method of IP Pool scaling will be added to CNS: CNS will watch Pods for its Node, and will request/release IPs immediately based on the number of Pods scheduled. The Batching behavior will be unchanged, and CNS will continue to request IPs in Batches $B$ based on the local IP usage.
+
```
Comment (nairashu) [inline docs/feature/subnet-scarcity/proposal.md:79]: Add in handling of the race condition that Ramiro brought up on Rahul's PR

**Review by nairashu — DISMISSED** [review]

File: docs/feature/subnet-scarcity/proposal.md (hunk 2 above)
Comment (rbtr) [reply docs/feature/subnet-scarcity/proposal.md:82]: noted, I have covered this in my local working draft and will include it in the next PR addition

File: docs/feature/subnet-scarcity/phase-1/1-subnetstate.md (hunk 1 above)
Comment (rsagasthya) [reply docs/feature/subnet-scarcity/phase-1/1-subnetstate.md:37]: The response codes are 400 for invalid networkId or subnetName, 500 in cases of error in retrieving the cache from controller. Success is 200.

**Review by rsagasthya — APPROVED** [review]

**Pull request merged**